File: my-folder/0649-dota2-senate/solution.java
class Solution {
    public String predictPartyVictory(String senate) {
        Queue<Integer> rad = new LinkedList<>();
        Queue<Integer> dir = new LinkedList<>();
        int n = senate.length();
        for(int i=0;i<n;i++){
            if(senate.charAt(i) == 'R')
                rad.offer(i);
            else
                dir.offer(i);
        }
        while(!rad.isEmpty() && !dir.isEmpty()){
            int v1 = rad.poll();
            int v2 = dir.poll();
            if(v1<v2)
                rad.offer(n+v1);
            else
                dir.offer(n+v2);
        }
        return (rad.isEmpty()) ? ("Dire") : ("Radiant");
    }
}

File: my-folder/0649-dota2-senate/solution.py
class Solution:
    def predictPartyVictory(self, senate: str) -> str:
        R, D = deque(), deque()
        n = len(senate)

        for idx, team in enumerate(senate):
            if team == 'R':
                R.append(idx)
            else:
                D.append(idx)

        while R and D:
            Rturn = R.popleft()
            Dturn = D.popleft()

            if Rturn < Dturn:
                R.append(Rturn + n)
            else:
                D.append(Dturn + n)

        return "Radiant" if R else "Dire"

# Time -> O(n)
# Space -> O(n)
        
